src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -405,13 +471,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            FreeDeconstructionVariables(variables);
+
+            return result;
+        }
+
+        /// <summary>
+        /// Prepares locals corresponding to the variables of the declaration.
+        /// The locals are kept in a tree which captures the nesting of variables.
+        /// Each local is either a simple local (when its type is known) or a deconstruction local pending inference.
+        /// The caller is responsible for releasing the nested ArrayBuilders.
+        /// </summary>
+        private ArrayBuilder<DeconstructionVariable> BindDeconstructionDeclarationVariables(VariableDeclarationSyntax node, TypeSyntax varSyntax, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Deconstruction != null);
+            SeparatedSyntaxList<VariableDeclarationSyntax> variables = node.Deconstruction.Variables;
+
+            // There are four cases for VariableDeclaration:
+            // - type and declarators are set, but deconstruction is null. This could represent `int x`, which is a single variable.
+            // - type is null, declarators are set, but deconstruction is null. This could represent `x`, which is a single variable.
+            // - type is set to 'var', declarators are null, and deconstruction is set. This could represent `var (...)`
+            // - type and declarators are null, but deconstruction is set. This could represent `(int x, ...)`
+
+            var localsBuilder = ArrayBuilder<DeconstructionVariable>.GetInstance(variables.Count);
+            foreach (var variable in variables)
+            {
+                TypeSyntax variableVarSyntax = varSyntax ?? variable.Type;
+
+                DeconstructionVariable local;
+                if (variable.Deconstruction == null)
+                {
+                    local = new DeconstructionVariable(BindDeconstructionDeclarationVariable(variable, variableVarSyntax, diagnostics));
+                }
+                else
+                {
+                    local = new DeconstructionVariable(BindDeconstructionDeclarationVariables(variable, variableVarSyntax, diagnostics));
+                }
+
+                localsBuilder.Add(local);
+            }
+
+            return localsBuilder;
+        }
+
+        /// <summary>
+        /// Returns a BoundLocal when the type was explicit, otherwise returns a DeconstructionLocalPendingInference.
+        /// </summary>
+        private BoundExpression BindDeconstructionDeclarationVariable(VariableDeclarationSyntax node, TypeSyntax varSyntax,DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Deconstruction == null);
+            Debug.Assert(varSyntax != null || node.Type != null);
+            Debug.Assert(node.Variables.Count == 1);
+
+            var declarator = node.Variables[0];
+            var typeSyntax = varSyntax ?? node.Type;
```

Comment:
Seems like a comment was lost. I removed the double-calculation and renamed the parameter to `closestTypeSyntax` (here and some other places).